### DISCUSSION TITLE
OutofTree Controller

### DISCUSSION THREAD

**Discussion by aasignacion** [bitcraze/discussions#1448] [Q&A]
I made a OOT code for the flapper drone.
I checked that for Flapper Drones and PID controller is forced. 
Because of the 4th line
in https://github.com/bitcraze/crazyflie-firmware/blob/master/configs/flapper_defconfig.

So I added this line CONFIG_CONTROLLER_OOT=y.  Then

make clean
make flapper_defconfig
make -j8

But I think when I added the CONFIG_CONTROLLER_OOT=y. line in flapper_defconfig, this error occurs:

/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: src/
built-in.o:(.rodata.controllerFunctions+0x60): undefined reference to `
controllerOutOfTreeInit'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: src/
built-in.o:(.rodata.controllerFunctions+0x64): undefined reference to `
controllerOutOfTreeTest'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: src/
built-in.o:(.rodata.controllerFunctions+0x68): undefined reference to `
controllerOutOfTree'
collect2: error: ld returned 1 exit status
make[1]: *** [tools/kbuild/Makefile.kbuild:570: firmware] Error 1
make[1]: Leaving directory '/home/crazyflie/Documents/crazyflie-firmware
/build'
make: *** [tools/kbuild/Makefile.kbuild:150: sub-make] Error 2

**It seems adding this line in any defconfig file [even for crazyflie 2.1], this error occurs.**

I already emailed and asked Matej about this and he said 
[From the error messages you are getting when you build your code, it seems
that the OutOfTree controller files/functions cannot be located during the
build process.
When I tried the OutOfTree functionality some time ago (for an estimator),
these were the files I had to add, maybe there is something missing on your
end?] - I checked the files and it matches all the files that Matej mentioned.

Could you help me to solve this problem?

**Comment by ataffanel** [comment]
Are you adding your own controller to the build system as well?

This error comes from the linker not finding the out of tree required functions. When you activate out of tree controllers, the firmware will simply add a controller with predefined function names: https://github.com/bitcraze/crazyflie-firmware/blob/ec6a7929d5b17fbffa82fecbfd5c33b27155171d/src/modules/src/controller/controller.c#L33-L35. You need to add these functions in a file that is compiled-in the Crazyflie firmware.

So you should double check that the file that implements the out of tree controller functions is actually being compiled and contains the right function names.
